File: server/src/main/java/br/com/diegosilva/automation/actors/Device.java
package br.com.diegosilva.automation.actors;

import akka.actor.Cancellable;
import akka.actor.typed.ActorRef;
import akka.actor.typed.ActorSystem;
import akka.actor.typed.Behavior;
import akka.actor.typed.PreRestart;
import akka.actor.typed.javadsl.AbstractBehavior;
import akka.actor.typed.javadsl.ActorContext;
import akka.actor.typed.javadsl.Behaviors;
import akka.actor.typed.javadsl.Receive;
import akka.cluster.sharding.typed.javadsl.ClusterSharding;
import akka.cluster.sharding.typed.javadsl.Entity;
import akka.cluster.sharding.typed.javadsl.EntityTypeKey;
import br.com.diegosilva.automation.CborSerializable;
import br.com.diegosilva.automation.dto.IOTMessage;
import br.com.diegosilva.automation.utils.SerialPortFactory;
import com.fasterxml.jackson.annotation.JsonCreator;
import jssc.SerialPort;
import jssc.SerialPortException;

import java.time.Duration;

public class Device extends AbstractBehavior<Device.Command> {

    private Cancellable cancellable;
    private final String deviceId;
    private SerialPort serialPort;


    public static final EntityTypeKey<Command> TypeKey =
            EntityTypeKey.create(Device.Command.class, "Device");

    public static void initSharding(ActorSystem<?> system) {
        ClusterSharding.get(system).init(Entity.of(TypeKey, entityContext ->
                Device.create(entityContext.getEntityId())
        ));
    }

    public static Behavior<Command> create(String deviceid) {
        return Behaviors.setup(context -> new Device(context, deviceid));
    }

    private Device(ActorContext<Command> context, String deviceId) throws SerialPortException {
        super(context);
        this.deviceId = deviceId;
        this.serialPort = SerialPortFactory.get(context.getSystem().settings().config().getString("serial.port"));
    }

    @Override
    public Receive<Command> createReceive() {
        return newReceiveBuilder()
                .onMessage(Process.class, (msg) -> {
                    getContext().getLog().info("Processando mensagem IOT {}", msg.message);
                    return this;
                })
                .onMessage(Send.class, (msg) -> {
                    getContext().getLog().info("Escrevendo mensagem na serial {}", msg.message.encode());
                    serialPort.writeString(msg.message.encode() + "\n");
                    if(msg.times < 3)
                        retry(Duration.ofSeconds(2), new Send(msg.message, msg.times+1, null));
                    if (msg.replyTo != null)
                        msg.replyTo.tell(new SendResponse("Mensagem enviada"));
                    return this;
                })
                .onSignal(
                        PreRestart.class,
                        signal -> {
                            getContext().getLog().debug("Reiniciando de recebimento de leituras");
                            this.serialPort = SerialPortFactory.get(getContext().getSystem().settings().config().getString("serial.port"));
                            return this;
                        })
                .build();
    }


    private void retry(Duration duration, Command cmd) {
        if (this.cancellable != null) {
            this.cancellable.cancel();
        }
        this.cancellable = getContext().scheduleOnce(duration, getContext().getSelf(), cmd);
    }


    public interface Command extends CborSerializable {
    }

    public interface Response extends CborSerializable {
    }

    public static class Process implements Command {
        final IOTMessage message;

        @JsonCreator
        public Process(IOTMessage message) {
            this.message = message;
        }
    }


    public static class Send implements Command {
        final IOTMessage message;
        final ActorRef<Response> replyTo;
        final int times;

        @JsonCreator
        public Send(IOTMessage message, int times, ActorRef<Response> replyTo) {
            this.message = message;
            this.replyTo = replyTo;
            this.times = times;
        }
    }

    public static class SendResponse implements Response {
        public final String message;

        @JsonCreator
        public SendResponse(String message) {
            this.message = message;
        }
    }
}

File: server_scl/src/main/scala/br/com/diegosilva/home/actors/Device.scala
package br.com.diegosilva.home.actors

import akka.actor.Cancellable
import akka.actor.typed.scaladsl.{AbstractBehavior, ActorContext, Behaviors}
import akka.actor.typed.{ActorRef, ActorSystem, Behavior, PreRestart, Signal}
import akka.cluster.sharding.typed.scaladsl.{ClusterSharding, Entity, EntityTypeKey}
import br.com.diegosilva.home.CborSerializable
import br.com.diegosilva.home.actors.Device.{Command, Processar, Send}
import br.com.diegosilva.home.dto.IOTMessage
import br.com.diegosilva.home.factory.SerialPortFactory
import jssc.SerialPort

import scala.concurrent.duration._

object Device {

  sealed trait Command extends CborSerializable

  sealed trait Response extends CborSerializable

  final case class Processar(message: IOTMessage) extends Command

  final case class Send(message: IOTMessage, replyTo: ActorRef[Response], times: Int) extends Command

  final case class SendResponse(message: String) extends Response


  val EntityKey: EntityTypeKey[Command] = EntityTypeKey[Command]("device")

  def init(system: ActorSystem[_]): Unit = {
    ClusterSharding(system).init(Entity(EntityKey) { entityContext =>
      create(entityContext.entityId)
    })
  }

  def create(deviceid: String): Behavior[Command] =
    Behaviors.setup(context => new Device(context, deviceid))

}

class Device(context: ActorContext[Command], val entityId: String) extends AbstractBehavior[Command](context) {

  private var cancellable: Cancellable = null
  private var serialPort: SerialPort = SerialPortFactory.get(context.system.settings.config.getString("serial.port"))

  override def onMessage(msg: Command): Behavior[Command] = {
    msg match {
      case Processar(message) => {
        context.log.info("Processando mensagem IOT {}", message)
        Behaviors.same
      }
      case Send(message, replyTo, times) => {
        if (cancellable != null)
          cancellable.cancel()
        context.log.info("Escrevendo mensagem na serial {}", message.encode)
        serialPort.writeString(message.encode + "\n")
        if (times < 3) retry(2.seconds, Device.Send(message, null, times + 1))
        if (replyTo != null) replyTo ! Device.SendResponse("Mensagem enviada")
        Behaviors.same
      }
    }
  }

  override def onSignal: PartialFunction[Signal, Behavior[Command]] = {
    case restart: PreRestart => {
      context.log.debug("Reiniciando de recebimento de leituras")
      this.serialPort = SerialPortFactory.get(context.system.settings.config.getString("serial.port"))
      Behaviors.same
    }
  }

  private def retry(duration: FiniteDuration, cmd: Device.Command): Unit = {
    if (this.cancellable != null) this.cancellable.cancel
    this.cancellable = context.scheduleOnce(duration, context.self, cmd)
  }
}